File: Raven.Client.Lightweight/Document/BulkInsertOperation.cs
using Raven.Abstractions.Util;
using Raven.Client.Connection.Async;
using System;
using System.Threading.Tasks;
using Raven.Abstractions.Data;
using Raven.Client.Changes;
using Raven.Client.Extensions;
using Raven.Json.Linq;

namespace Raven.Client.Document
{
    public class BulkInsertOperation : IDisposable
    {
        public Guid OperationId
        {
            get
            {
                return Operation.OperationId;
            }
        }

        private readonly IDocumentStore documentStore;
        private readonly GenerateEntityIdOnTheClient generateEntityIdOnTheClient;
        protected internal ILowLevelBulkInsertOperation Operation { get; set; }
        public IAsyncDatabaseCommands DatabaseCommands { get; private set; }
        private readonly EntityToJson entityToJson;

        public delegate void BeforeEntityInsert(string id, RavenJObject data, RavenJObject metadata);

        public event BeforeEntityInsert OnBeforeEntityInsert = delegate { };

        public bool IsAborted
        {
            get { return Operation.IsAborted; }
        }

        public void Abort()
        {
            Operation.Abort();
        }

        public event Action<string> Report
        {
            add { Operation.Report += value; }
            remove { Operation.Report -= value; }
        }

        public BulkInsertOperation(string database, IDocumentStore documentStore, DocumentSessionListeners listeners, BulkInsertOptions options, IDatabaseChanges changes)
        {
            this.documentStore = documentStore;

            database = database ?? MultiDatabase.GetDatabaseName(documentStore.Url);

            // Fitzchak: Should not be ever null because of the above code, please refactor this.
            DatabaseCommands = database == null
                ? documentStore.AsyncDatabaseCommands.ForSystemDatabase()
                : documentStore.AsyncDatabaseCommands.ForDatabase(database);

            generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(documentStore.Conventions, entity =>
                AsyncHelpers.RunSync(() => documentStore.Conventions.GenerateDocumentKeyAsync(database, DatabaseCommands, entity)));

            Operation = GetBulkInsertOperation(options, DatabaseCommands, changes);
            entityToJson = new EntityToJson(documentStore, listeners);
        }

        protected virtual ILowLevelBulkInsertOperation GetBulkInsertOperation(BulkInsertOptions options, IAsyncDatabaseCommands commands, IDatabaseChanges changes)
        {
            return commands.GetBulkInsertOperation(options, changes);
        }

        public Task DisposeAsync()
        {
            return Operation.DisposeAsync();
        }

        public Task WaitForLastTaskToFinish()
        {
            return Operation.WaitForLastTaskToFinish();
        }

        public void Dispose()
        {
            Operation.Dispose();
        }

        public string Store(object entity)
        {
            var id = GetId(entity);
            Store(entity, id);
            return id;
        }

        public async Task<string> StoreAsync(object entity)
        {
            var id = GetId(entity);
            await StoreAsync(entity, id).ConfigureAwait(false);
            return id;
        }

        public void Store(object entity, string id)
        {
            if (Operation.IsAborted)
                throw new InvalidOperationException("Bulk insert has been aborted or the operation was timed out");

            var metadata = new RavenJObject();

            var tag = documentStore.Conventions.GetDynamicTagName(entity);
            if (tag != null)
                metadata.Add(Constants.RavenEntityName, tag);

            var data = entityToJson.ConvertEntityToJson(id, entity, metadata);

            OnBeforeEntityInsert(id, data, metadata);

            Operation.Write(id, metadata, data);
        }

        public Task StoreAsync(object entity, string id)
        {
            if (Operation.IsAborted)
                throw new InvalidOperationException("Bulk insert has been aborted or the operation was timed out");

            var metadata = new RavenJObject();

            var tag = documentStore.Conventions.GetDynamicTagName(entity);
            if (tag != null)
                metadata.Add(Constants.RavenEntityName, tag);

            var data = entityToJson.ConvertEntityToJson(id, entity, metadata);

            OnBeforeEntityInsert(id, data, metadata);

            return Operation.WriteAsync(id, metadata, data);
        }

        public void Store(RavenJObject document, RavenJObject metadata, string id, int? dataSize = null)
        {
            OnBeforeEntityInsert(id, document, metadata);

            Operation.Write(id, metadata, document, dataSize);
        }

        public Task StoreAsync(RavenJObject document, RavenJObject metadata, string id, int? dataSize = null)
        {
            OnBeforeEntityInsert(id, document, metadata);

            return Operation.WriteAsync(id, metadata, document, dataSize);
        }

        private string GetId(object entity)
        {
            string id;
            if (generateEntityIdOnTheClient.TryGetIdFromInstance(entity, out id) == false)
            {
                id = generateEntityIdOnTheClient.GenerateDocumentKeyForStorage(entity);
            }
            return id;
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/document/BulkInsertOperation.java
package net.ravendb.client.document;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.closure.Function1;
import net.ravendb.abstractions.data.BulkInsertOptions;
import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.changes.IDatabaseChanges;
import net.ravendb.client.connection.IDatabaseCommands;
import net.ravendb.client.extensions.MultiDatabase;



public class BulkInsertOperation implements AutoCloseable {

  public static interface BeforeEntityInsert {
    public void apply(String id, RavenJObject data, RavenJObject metadata);
  }

  private final IDocumentStore documentStore;
  private final GenerateEntityIdOnTheClient generateEntityIdOnTheClient;
  private final ILowLevelBulkInsertOperation operation;
  private IDatabaseCommands databaseCommands;

  private final EntityToJson entityToJson;

  private Set<BeforeEntityInsert> onBeforeEntityInsert = new LinkedHashSet<>();

  public Action1<String> getReport() {
    return operation.getReport();
  }


  public void setReport(Action1<String> report) {
    operation.setReport(report);
  }

  public void addOnBeforeEntityInsert(BeforeEntityInsert action) {
    onBeforeEntityInsert.add(action);
  }

  public void removeOnBeforeEntityInsert(BeforeEntityInsert action) {
    onBeforeEntityInsert.remove(action);
  }

  public IDatabaseCommands getDatabaseCommands() {
    return databaseCommands;
  }

  public UUID getOperationId() {
    return operation.getOperationId();
  }

  public BulkInsertOperation(final String database, final IDocumentStore documentStore, DocumentSessionListeners listeners, BulkInsertOptions options, IDatabaseChanges changes) {
    this.documentStore = documentStore;
    final String finalDatabase = (database != null) ? database : MultiDatabase.getDatabaseName(documentStore.getUrl());
    // Fitzchak: Should not be ever null because of the above code, please refactor this.
    databaseCommands = finalDatabase == null ? documentStore.getDatabaseCommands().forSystemDatabase()
      :documentStore.getDatabaseCommands().forDatabase(finalDatabase);
    generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(documentStore, new Function1<Object, String>() {
      @Override
      public String apply(Object entity) {
        return documentStore.getConventions().generateDocumentKey(finalDatabase, getDatabaseCommands(), entity);
      }
    });
    operation = databaseCommands.getBulkInsertOperation(options, changes);
    entityToJson = new EntityToJson(documentStore, listeners);
  }

  @Override
  public void close() throws Exception {
    operation.close();
  }

  public String store(Object entity) {
    String id = getId(entity);
    store(entity, id);
    return id;
  }

  public void store(Object entity, String id) {
    RavenJObject metadata = new RavenJObject();

    String tag = documentStore.getConventions().getTypeTagName(entity.getClass());
    if (tag != null) {
      metadata.add(Constants.RAVEN_ENTITY_NAME, tag);
    }
    RavenJObject data = entityToJson.convertEntityToJson(id, entity, metadata);

    onBeforeEntityInsert(id, data, metadata);

    operation.write(id, metadata, data);
  }

  private void onBeforeEntityInsert(String id, RavenJObject data, RavenJObject metadata) {
    for (BeforeEntityInsert event: onBeforeEntityInsert) {
      event.apply(id, data, metadata);
    }
  }

  public void store(RavenJObject document, RavenJObject metadata, String id) {
    onBeforeEntityInsert(id, document, metadata);

    operation.write(id, metadata, document);
  }

  public String getId(Object entity) {
    Reference<String> idRef = new Reference<>();

    if (generateEntityIdOnTheClient.tryGetIdFromInstance(entity, idRef)) {
      idRef.value = generateEntityIdOnTheClient.generateDocumentKeyForStorage(entity);
    } else {
      idRef.value = generateEntityIdOnTheClient.generateDocumentKeyForStorage(entity);
      generateEntityIdOnTheClient.trySetIdentity(entity, idRef.value);
    }
    return idRef.value;
  }
}
